cpp/src/arrow/ipc/metadata_internal.cc

```diff
@@ -1349,17 +1349,23 @@ Status GetTensorMetadata(const Buffer& metadata, std::shared_ptr<DataType>* type
     return Status::IOError("Header-type of flatbuffer-encoded Message is not Tensor.");
   }
 
-  int ndim = static_cast<int>(tensor->shape()->size());
+  const auto ndim = tensor->shape()->size();
+  using ndim_type = std::remove_const<decltype(ndim)>::type;
```

Comment:
Simple `decltype(ndim)` is modified by `const` so we need to use `std::remove_const` here.